### PR TITLE
feat(modal): update modal options

### PR DESCRIPTION
**Pull request by SophieBatumbo**
Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) and [DEVELOPER.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/DEVELOPER.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [x] added/updated any applicable demos.

### CODE REVIEW

File: src/modal/modal-ref.ts [hunk 1]
```diff
@@ -40,6 +67,55 @@ export class NgbModalRef {
 	private _hidden = new Subject<void>();
 	private _resolve: (result?: any) => void;
 	private _reject: (reason?: any) => void;
+	private _windowAttributes = [
```
Comment (maxokorokov) [inline src/modal/modal-ref.ts:70]: Let's make this not private to `ModalRef`, but just a global `const WINDOW_ATTRIBUTES = [...]`

File: src/modal/modal.scss [hunk 2]
```diff
@@ -5,3 +5,9 @@ ngb-modal-window {
 		overflow: hidden;
 	}
 }
+
+.foo {
```
Comment (maxokorokov) [inline src/modal/modal.scss:9]: Please remove

File: src/modal/modal.spec.ts [hunk 3]
```diff
@@ -633,6 +727,95 @@ describe('ngb-modal', () => {
...
+		// The usage of private property does not seem to be a clean access way isn't it ?
+		// That's why something like : expect(modalInstance).toHaveOption('ariaDescibedBy').toBe('blabla');
+		// or a public getter like   : expect(modalInstance.getOptions()['ariaDescibedBy']).toBe('blabla');
+		// did seem to be a great idea according to me
+		describe('update of ariaDescibedBy options ', () => {
+			it('should render modals with the correct ariaDescibedBy after update', () => {
+				const modalInstance = fixture.componentInstance.open('foo', { ariaDescribedBy: 'aLabel' });
+				fixture.detectChanges();
+				expect(fixture.nativeElement).toHaveModal('foo');
+				expect(modalInstance._windowCmptRef.instance.ariaDescribedBy).toBe('aLabel');
```
Comment (maxokorokov) [inline src/modal/modal.spec.ts:773]: as discussed you should check attributes in the resulting DOM here

File: src/modal/modal-ref.ts [hunk 4]
```diff
@@ -29,8 +31,33 @@ export class NgbActiveModal {
 	 * The `NgbModalRef.result` promise will be rejected with the provided value.
 	 */
 	dismiss(reason?: any): void {}
+
+	/**
+	 * Update the modal with `options` value. FAUX OMG A MODIF
```
Comment (maxokorokov) [inline src/modal/modal-ref.ts:36]: Please fix the documentation

File: demo/src/app/components/modal/demos/component/modal-component.ts [hunk 5]
```diff
@@ -10,9 +10,33 @@ import { NgbActiveModal, NgbModal } from '@ng-bootstrap/ng-bootstrap';
 			<button type="button" class="btn-close" aria-label="Close" (click)="activeModal.dismiss('Cross click')"></button>
 		</div>
 		<div class="modal-body">
-			<p>Hello, {{ name }}!</p>
+			<pre>Hello, {{ name }}!</pre>
```
Comment (maxokorokov) [inline demo/src/app/components/modal/demos/component/modal-component.ts:13]: why change markup?

**Review by maxokorokov — CHANGES_REQUESTED** [review]
Hey, thanks for the PR!

Please make sure that:
- commits are squashed or at least meaningful
- tests pass (it's red now)
- no 'development' code left in the PR (like `.foo` class in styles)
- no commented out code left

Left some comments inline!

Cheers!

File: src/modal/modal-ref.ts [hunk 6]
```diff
@@ -40,6 +67,55 @@ export class NgbModalRef {
...
+
+	private _applyBackdropOptions(backdropInstance: NgbModalBackdrop, options: NgbModalOptions): void {
+		this._backdropAttributes.forEach((optionName: string) => {
+			if (isDefined(options[optionName])) {
+				backdropInstance[optionName] = options[optionName];
+			}
+		});
+	}
+
+	update(
```
Comment (maxokorokov) [inline src/modal/modal-ref.ts:101]: Documentation missing for this method. Once you add comment, it should be documented here (https://ng-bootstrap.github.io/#/components/modal/api#NgbModalRef) alongside the `close` and `dismiss` methods

File: src/modal/modal-ref.ts [hunk 7]
```diff
@@ -40,6 +67,55 @@ export class NgbModalRef {
...
+		'backdrop',
+		'centered',
+		'fullscreen',
+		'keyboard',
+		'scrollable',
+		'size',
+		'windowClass',
+		'modalDialogClass',
+	];
+	private _backdropAttributes = ['animation', 'backdropClass'];
```
[the remark this answers is not in the document]
Comment (SophieBatumbo) [reply src/modal/modal-ref.ts:83]: Same reply as for` _windowAttributes`

File: src/modal/modal-ref.ts [hunk 8]
```diff
@@ -29,8 +31,33 @@ export class NgbActiveModal {
...
+
+	/**
+	 * Update the modal with `options` value. FAUX OMG A MODIF
+	 *
+	 * The `NgbModalRef.options` promise will be resolved with the provided value.
+	 */
+	update(options: NgbModalOptions): void {}
+}
+
+/*export interface NgbModalUpdatableOptions {
```
[the remark this answers is not in the document]
Comment (SophieBatumbo) [reply src/modal/modal-ref.ts:43]: Done

File: src/modal/modal-stack.ts [hunk 9]
```diff
@@ -131,13 +117,30 @@ export class NgbModalStack {
...
+
+		ngbModalRef.update(options);
+		//this._applyWindowOptions(windowCmptRef.instance, options);
 		if (this._modalRefs.length === 1) {
 			renderer.addClass(this._document.body, 'modal-open');
 		}
 
 		if (backdropCmptRef && backdropCmptRef.instance) {
-			this._applyBackdropOptions(backdropCmptRef.instance, options);
+			//this._applyBackdropOptions(backdropCmptRef.instance, options);
```
[the remark this answers is not in the document]
Comment (SophieBatumbo) [reply src/modal/modal-stack.ts:143]: Deleted

File: src/modal/modal.scss (hunk 2 above)
Comment (SophieBatumbo) [reply src/modal/modal.scss:9]: Deleted

File: src/modal/modal.spec.ts (hunk 3 above)
Comment (SophieBatumbo) [reply src/modal/modal.spec.ts:773]: Done

File: demo/src/app/components/modal/demos/component/modal-component.ts [hunk 10]
```diff
@@ -10,9 +10,33 @@ import { NgbActiveModal, NgbModal } from '@ng-bootstrap/ng-bootstrap';
...
 		</div>
 		<div class="modal-footer">
+			<button type="button" class="btn btn-outline-dark" (click)="activeModal.update({ size: 'xl' })">
```
[the remark this answers is not in the document]
Comment (SophieBatumbo) [reply demo/src/app/components/modal/demos/component/modal-component.ts:16]: Done

File: src/modal/modal-ref.ts (hunk 6 above)
Comment (SophieBatumbo) [reply src/modal/modal-ref.ts:101]: Done

File: src/modal/modal-ref.ts [hunk 11]
```diff
@@ -5,17 +5,36 @@ import { takeUntil } from 'rxjs/operators';
 
 import { NgbModalBackdrop } from './modal-backdrop';
 import { NgbModalWindow } from './modal-window';
+import { NgbModalOptions } from './modal-config';
+import { isDefined } from '../util/util';
 
 import { ContentRef } from '../util/popup';
 import { isPromise } from '../util/util';
 
+export type NgbUpdatableModalOptions = Pick<
```
Comment (maxokorokov) [inline src/modal/modal-ref.ts:14]: Could we rename it to `NgbModalUpdatableOptions` for consistency with `NgbModalOptions` ?

File: src/modal/modal-ref.ts [hunk 12]
```diff
@@ -31,6 +50,23 @@ export class NgbActiveModal {
 	dismiss(reason?: any): void {}
 }
 
+class ModalAttributes {
```
Comment (maxokorokov) [inline src/modal/modal-ref.ts:53]: why have the whole class for this?

Just `const WINDOW_ATTRIBUTES: string[] = [...]` would be enough

**Review by maxokorokov — CHANGES_REQUESTED** [review]
Hey @SophieBatumbo, we're almost good here, thanks!

Just three small changes:
- [ ] Please rename to `NgbModalUpdatableOptions`
- [ ] Please remove static class for window/backdrop attributes, just `const` will have the same effect
- [ ] Add `NgbModalUpdatableOptions` as an export to the public API (here → https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/index.ts#L102, and here → https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/modal/modal.module.ts#L6). Otherwise it will not be available when importing from `@ng-bootstrap/ng-bootstrap`.

We're good after this, I'll merge the PR.

Cheers,
Max

File: src/modal/modal-ref.ts (hunk 12 above)
Comment (SophieBatumbo) [reply src/modal/modal-ref.ts:53]: As const can not be used for a class member you'll see `readonly` , done

File: src/modal/modal-ref.ts (hunk 11 above)
Comment (SophieBatumbo) [reply src/modal/modal-ref.ts:14]: Done

**Comment by SophieBatumbo** [comment]
- [ ]  Please rename to NgbModalUpdatableOptions

- [ ] Please remove static class for window/backdrop attributes, just const will have the same effect

- [ ]  Add NgbModalUpdatableOptions as an export to the public API (here → https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/index.ts#L102, and here → https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/modal/modal.module.ts#L6). Otherwise it will not be available when importing from @ng-bootstrap/ng-bootstrap.

Everything is done, the comment documentation is added too for `NgbModalUpdatableOptions`

**Review by maxokorokov — APPROVED** [review]
Thanks, I'll merge this in `14.2` release

**Pull request merged**